src/aind_registration_evaluation/util/extract_roi.py

```diff
@@ -0,0 +1,291 @@
+                                    'addedDist': addedDist}}
+    '''
+    distances = {}
+    for p1 in props1:
+        for p2 in props2:
+            dist = np.linalg.norm(np.array(p1.centroid) - np.array(p2.centroid))
+            if dist < distanceThreshold:
+                addedDist = 0
+                while dist in distances:
+                    dist += 0.0001
+                    addedDist += 0.0001
+                                # distances[dist] = {'Label1': p1.label, 'Centroid1': p1.centroid, 'Bbox1': p1.bbox,
+                #                     'Label2': p2.label, 'Centroid2': p2.centroid, 'Bbox2': p2.bbox,
+                #                     'addedDist': addedDist}
+                distances[dist] = {'R1': p1, 'R2': p2,'addedDist': addedDist}
+    sortedDistances = {k: v for k, v in sorted(distances.items(), key=lambda item: int(item[0]) + item[1]['addedDist'])}
+    p1Used = set()
+    p2Used = set()
+    resultDistances = {}
+    for k,v in sortedDistances.items():
+        if v['R1'].label in p1Used or v['R2'].label in p2Used: continue
+        resultDistances[k] = v
+        p1Used.add(v['R1'].label)
+        p2Used.add(v['R2'].label)
+    return resultDistances
+
+def create_patches(distances, image_shape, overlapThreshold):
+    '''
+        - Creates a patch around each matched centroids by adding 3 pixels around the max coordinates of the patches (limited by image bounds).
+        - Merges the patches that show more overlap than the overlap threshold, to avoid duplicate evaluations on the same area.
+
+        Parameters
+        ------------------------
+       distances: Dictionary of shape {distance:{'R1': p1, 'R2': p2,'addedDist': addedDist}}
+       
+       image_shape: shape of the two images must be the same. Either can be passed in the form Tuple(z,y,x)
+
+       overlapThreshold: volume overlap theshold above which two patches are merged to their max coordinates (float)
+
+        Returns
+        ------------------------
+        coordinates of the merged_patches (z_min, y_min, x_min, z_max, y_max, x_max)
+    '''
+    patches = []
+    for distance, info in distances.items():
+        bbox1 = info['R1'].bbox
+        bbox2 = info['R2'].bbox
+        z_min = 0 if min(bbox1[0],bbox2[0])<4 else (min(bbox1[0],bbox2[0])-3) 
+        y_min = 0 if min(bbox1[1],bbox2[1])<4 else (min(bbox1[1],bbox2[1])-3)
+        x_min = 0 if min(bbox1[2],bbox2[2])<4 else (min(bbox1[2],bbox2[2])-3)
+        z_max = image_shape[0] if max(bbox1[3],bbox2[3])>(image_shape[0]-4) else (max(bbox1[3],bbox2[3])+3)
+        y_max = image_shape[1] if max(bbox1[4],bbox2[4])>(image_shape[1]-4) else (max(bbox1[4],bbox2[4])+3)
+        x_max = image_shape[2] if max(bbox1[5],bbox2[5])>(image_shape[2]-4) else (max(bbox1[5],bbox2[5])+3)
+        patches.append((z_min, y_min, x_min, z_max, y_max, x_max))
+    
+    merged_patches = []
+    while patches:
+        z_min, y_min, x_min, z_max, y_max, x_max = patches.pop(0)
```

Comment:
This block of code is super dense, please break. Also, the while must be used only when you don't know how many times you will execute the loop. In this case, I see patches is a list which has a finite number of elements that you're adding.

Reply:
It looks it is some optimization you're doing? If so, please ignore my comment. But break the code, or explain it a little better :)